crates/ra_hir_ty/src/expr.rs

```diff
@@ -52,15 +53,89 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         for e in body.exprs.iter() {
             if let (id, Expr::RecordLit { path, fields, spread }) = e {
                 self.validate_record_literal(id, path, fields, *spread, db);
+            } else if let (id, Expr::Match { expr, arms }) = e {
+                self.validate_match(id, *expr, arms, db, self.infer.clone());
             }
         }
 
         let body_expr = &body[body.body_expr];
-        if let Expr::Block { statements: _, tail: Some(t) } = body_expr {
+        if let Expr::Block { tail: Some(t), .. } = body_expr {
             self.validate_results_in_tail_expr(body.body_expr, *t, db);
         }
     }
 
+    fn validate_match(
+        &mut self,
+        id: ExprId,
+        match_expr: ExprId,
+        arms: &[MatchArm],
+        db: &dyn HirDatabase,
+        infer: Arc<InferenceResult>,
+    ) {
+        let (body, source_map): (Arc<Body>, Arc<BodySourceMap>) =
+            db.body_with_source_map(self.func.into());
+
+        let match_expr_ty = match infer.type_of_expr.get(match_expr) {
+            Some(ty) => ty,
+            // If we can't resolve the type of the match expression
+            // we cannot perform exhaustiveness checks.
+            None => return,
+        };
+
+        let cx = MatchCheckCtx { body, infer: infer.clone(), db };
+        let pats = arms.iter().map(|arm| arm.pat);
+
+        let mut seen = Matrix::empty();
+        for pat in pats {
+            // We skip any patterns whose type we cannot resolve.
+            if let Some(pat_ty) = infer.type_of_pat.get(pat) {
+                // We only include patterns whose type matches the type
+                // of the match expression. If we had a InvalidMatchArmPattern
+                // diagnostic or similar we could raise that in an else
+                // block here.
+                //
+                // When comparing the types, we also have to consider that rustc
+                // will automatically de-reference the match expression type if
+                // necessary.
+                if pat_ty == match_expr_ty
+                    || match_expr_ty
+                        .as_reference()
+                        .map(|(match_expr_ty, _)| match_expr_ty == pat_ty)
+                        .unwrap_or(false)
```

Comment:
:thinking: it'd be better to check if the type checker reported type mismatches, though you'd have to walk the whole pattern to make sure no subpattern has a mismatch. But maybe this is enough right now (though I'd add a FIXME at least).

Alternatively, if we did the `is_useful` check for each pattern, we could just ignore patterns where it returns an error...

Reply:
> though you'd have to walk the whole pattern to make sure no subpattern has a mismatch

It seems to me like the current implementaiton does a "deep" type check, checking the types of all nested patterns. Is this true? If so, what is the advantage of using the type checker? Not to say I think we shouldn't use it here, I am just not familiar so I am trying to understand the trade-offs.

> Alternatively, if we did the is_useful check for each pattern, we could just ignore patterns where it returns an error...

I'm not sure what you mean here. I believe the current implementation skips patterns which don't type check.

Reply:
> I think the current implementaiton does a "deep" type check, checking the types of all nested patterns. Is this true? If so, what is the advantage of using the type checker? Not to say I think we shouldn't use it here, I am just not familiar so I am trying to understand the trade-offs.

Just as a general principle, comparing types is what the type checker does, and replicating that elsewhere is likely to get some edge cases wrong. I haven't thought deeply about this though. But you already have to replicate part of the match ergonomics logic, and are doing it only for one layer of references -- the match expression could be a `&&Something`, for example.

> I'm not sure what you mean here. I believe the current implementation skips patterns which don't type check?

Yeah, I meant this as an alternative to the current implementation.

Reply:
Got it, thanks for the explanation. I've added a fixme here for now, but I'm happy to look into this in a follow up MR if you can point me in the direction of the type checker :+1: